app_dart/lib/src/request_handlers/push_benchmark_to_center.dart

```diff
@@ -0,0 +1,144 @@
+import '../request_handling/body.dart';
+import '../request_handling/request_handler.dart';
+import '../service/datastore.dart';
+
+/// Pushes devicelab benchmark results to Metrics Center storage.
+///
+/// Metrics Center aggregates benchmark data in the unified [MetricPoint] format
+/// and sends them to performance dashboards.
+@immutable
+class PushBenchmarkToCenter extends RequestHandler<Body> {
+  PushBenchmarkToCenter(
+    Config config, {
+    @visibleForTesting Logging logger,
+    @visibleForTesting DatastoreService datastoreService,
+    @visibleForTesting FlutterDestinationProvider flutterDestinationProvider,
+  })  : logger = logger ?? Providers.serviceScopeLogger(),
+        datastoreService =
+            datastoreService ?? DatastoreService.defaultProvider(config.db),
+        flutterDestinationProvider =
+            flutterDestinationProvider ?? defaultFlutterDestinationProvider,
+        super(config: config);
+
+  final Logging logger;
+  final DatastoreService datastoreService;
+  final FlutterDestinationProvider flutterDestinationProvider;
+
+  @override
+  Future<Body> get() async {
+    final int sinceTimeMillis =
+        parseSinceTimeMillis(request.uri.queryParameters);
+    if (sinceTimeMillis == null) {
+      logger.error('The sinceTimeMillis parameter is missing');
+      return Body.empty;
+    }
+
+    final List<TimeSeriesValue> timeSeriesValues =
+        await readTimeSeriesValue(sinceTimeMillis);
+    final Map<String, TimeSeries> timeSeriesMap = await readTimeSeriesMap();
+
+    final List<BenchmarkMetricPoint> points =
+        await transform(timeSeriesMap, timeSeriesValues);
+    logger.debug('Transformed ${points.length} metric points.');
+    logger.debug('The first one is ${points.first}');
+
+    final mc.FlutterDestination destination =
+        await flutterDestinationProvider(config);
+    await destination.update(points);
+    return Body.empty;
+  }
+
+  @visibleForTesting
+  int parseSinceTimeMillis(Map<String, String> queryParameters) {
+    final String s = queryParameters['sinceTimeMillis'] ?? '';
+    return int.tryParse(s);
+  }
+
+  /// Returns a map containing all time series keyed by its id.
+  Future<Map<String, TimeSeries>> readTimeSeriesMap() async {
+    final Map<String, TimeSeries> map = <String, TimeSeries>{};
+    await for (TimeSeries ts in datastoreService.db.query<TimeSeries>().run()) {
+      // The id is string typed, see the defination of [TimeSeries].
+      map[ts.id as String] = ts;
+    }
+    return map;
+  }
+
+  /// Reads benchmark records with [createTimestamp] later than [sinceMillis].
+  /// And limits the result size to no larger than [batchSize].
+  Future<List<TimeSeriesValue>> readTimeSeriesValue(int sinceMillis,
+      {int batchSize = 1000}) async {
+    final Query<TimeSeriesValue> query =
+        datastoreService.db.query<TimeSeriesValue>()
+          ..filter('createTimestamp >', sinceMillis)
+          ..limit(batchSize)
```

Comment:
Do you think we'll need a more sophisticated batch size limit like https://github.com/liyuqian/metrics_center/blob/master/lib/src/flutter/center.dart#L121 ? Notice that we're returning points that are strictly later than a timestamp, and we would miss some points if many points have the same timestamp, and we cut the batch in the middle.